### ISSUE TITLE
Cross-compiling from macOS to iOS arm64 Undefined symbol: ___aarch64_ldadd8_relax

### ISSUE THREAD

**Issue by nmgwddj** [awslabs/aws-c-common#902]
compiler version: apple-clang 13.1
build system: macOS 12.3.1 arm64 M1
build target system: iOS 9.3 arm64

when I link aws-c-common library, I got this errors: 
```
Undefined symbols for architecture arm64:
  "___aarch64_ldadd8_relax", referenced from:
      _aws_atomic_fetch_add_explicit in libaws-c-common.a(ref_count.c.o)
      _aws_atomic_fetch_sub_explicit in libaws-c-common.a(ref_count.c.o)
  "___aarch64_ldadd8_acq", referenced from:
      _aws_atomic_fetch_add_explicit in libaws-c-common.a(ref_count.c.o)
      _aws_atomic_fetch_sub_explicit in libaws-c-common.a(ref_count.c.o)
  "___aarch64_ldadd8_rel", referenced from:
      _aws_atomic_fetch_add_explicit in libaws-c-common.a(ref_count.c.o)
      _aws_atomic_fetch_sub_explicit in libaws-c-common.a(ref_count.c.o)
  "___aarch64_ldadd8_acq_rel", referenced from:
      _aws_atomic_fetch_add_explicit in libaws-c-common.a(ref_count.c.o)
      _aws_atomic_fetch_sub_explicit in libaws-c-common.a(ref_count.c.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by graebm** [comment]
can you please the exact commands you're using to configure and build? And the output of the configure step?

**Comment by nmgwddj** [comment]
I use conan package to import aws-c-common, please see this link: https://github.com/conan-io/conan-center-index/issues/10477

**Comment by graebm** [comment]
I'm not able to reproduce the error, it builds OK for me (once I figure out how to set up that Xcode DEVELOPER_TEAM thing). I've tried building from both an Intel and M1 mac.

Here's what I did:
- installed conan (I didn't have it before)
- created a project like you described in that ticket
  - Here's a [zip file](https://github.com/awslabs/aws-c-common/files/8616854/conanios.zip) of what I made
  - Obviously, the `XCODE_ATTRIBUTE_DEVELOPMENT_TEAM` line in CMakeListst.txt needs be be changed
  - the `main.cc` file is just an empty main function. I'm not sure if that affects what gets linked into the final application.
- ran cmake configure and cmake build with the exact same commands described in that other ticket

Does the error happen for you if you nuke your `~/.conan/data` cache and build that new project?

**Comment by graebm** [comment]
Here are logs of my cmake configure, and cmake build commands, in case that helps.

[build_x86_64.txt](https://github.com/awslabs/aws-c-common/files/8617228/build_x86_64.txt)
[config_x86_64.txt](https://github.com/awslabs/aws-c-common/files/8617229/config_x86_64.txt)

These were logs were from an Intel mac. Before I ran this, I even edited my default conan profile to have the same CFLAGS and CXXFLAGS described in that ticket.

**Comment by nmgwddj** [comment]
> 这是我的 cmake 配置和 cmake 构建命令的日志，以防万一。
> 
> [build_x86_64.txt ](https://github.com/awslabs/aws-c-common/files/8617228/build_x86_64.txt) [config_x86_64.txt](https://github.com/awslabs/aws-c-common/files/8617229/config_x86_64.txt)
> 
> 这些是来自英特尔 mac 的日志。在我运行它之前，我什至编辑了我的默认柯南配置文件，使其具有该票证中描述的相同 CFLAGS 和 CXXFLAGS。

Thank you very much for your patient reply, I will make a simple project to reproduce the problem. Also please help to check thank you.

**Comment by nmgwddj** [comment]
https://github.com/nmgwddj/conan-cross-platform-build/tree/develop

This is a very simple example project to reproduce the problem we mentioned. The code is in the develop branch, and you can reproduce the problem with the following commands and steps:

```bash
cmake -Bbuild -GXcode \
	-DCMAKE_BUILD_TYPE=Debug \
	-DCMAKE_SYSTEM_NAME=iOS \
	-DCMAKE_OSX_DEPLOYMENT_TARGET=9.3 \
	-DCMAKE_OSX_ARCHITECTURES=arm64 \
	-DCONAN_PROFILE_BUILD=default \
	-DCONAN_PROFILE_HOST=$(pwd)/.profiles/ios-arm64-iphones-clang13
```

After cmake configuration, you may need to manually modify which certificate is used for signing. Then execute the following command to compile:

```bash
cmake —build build —config Debug
```

You will got some errors like this:

```
Undefined symbols for architecture arm64:
  "___aarch64_ldadd8_relax", referenced from:
      _aws_atomic_fetch_add_explicit in libaws-c-common.a(ref_count.c.o)
      _aws_atomic_fetch_sub_explicit in libaws-c-common.a(ref_count.c.o)
  "___aarch64_ldadd8_acq", referenced from:
      _aws_atomic_fetch_add_explicit in libaws-c-common.a(ref_count.c.o)
      _aws_atomic_fetch_sub_explicit in libaws-c-common.a(ref_count.c.o)
  "___aarch64_ldadd8_rel", referenced from:
      _aws_atomic_fetch_add_explicit in libaws-c-common.a(ref_count.c.o)
      _aws_atomic_fetch_sub_explicit in libaws-c-common.a(ref_count.c.o)
  "___aarch64_ldadd8_acq_rel", referenced from:
      _aws_atomic_fetch_add_explicit in libaws-c-common.a(ref_count.c.o)
      _aws_atomic_fetch_sub_explicit in libaws-c-common.a(ref_count.c.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by graebm** [comment]
Could you share the FULL logs from the config and build steps? Delete the conan cache at `~/.conan.data` beforehand, so that the logs show aws-c-common getting configured and built.

You can drag & drop a log file into the comment box and Github will upload them as files.

**Comment by nmgwddj** [comment]
[cmake configure](https://github.com/awslabs/aws-c-common/files/8644279/1.txt)
[cmake build](https://github.com/awslabs/aws-c-common/files/8644280/2.txt)

**Comment by graebm** [comment]
Thanks. I was actually able to reproduce it right before you posted those. Still investigating what is up...

**Comment by graebm** [comment]
try removing the `-fembed-bitcode` flag from these lines in your `.profiles/ios-arm64-iphones-clang13` file:
> CFLAGS=-fembed-bitcode -fvisibility=hidden -fvisibility-inlines-hidden
> CXXFLAGS=-fembed-bitcode -fvisibility=hidden -fvisibility-inlines-hidden

I got the error to reproduce in a simpler toy project that doesn't use conan. It just links against aws-c-common and calls some functions. I just figured out that the `-fembed-bitcode` flag is what breaks things, so sharing with you now. I'll keep digging into this next week to see if we can improve our CMake scripts.
